sw/device/silicon_creator/lib/sigverify.c

```diff
@@ -34,42 +105,65 @@
 static rom_error_t sigverify_padding_and_digest_check(
-    const sigverify_rsa_buffer_t *enc_msg, const hmac_digest_t *act_digest) {
-  const uint32_t *enc_msg_ptr = enc_msg->data;
+    sigverify_rsa_buffer_t *enc_msg, const hmac_digest_t *act_digest,
+    uint32_t *flash_exec) {
+  uint32_t *enc_msg_ptr = enc_msg->data;
 
-  if (memcmp(enc_msg_ptr, act_digest->digest, sizeof(act_digest->digest)) !=
-      0) {
-    return kErrorSigverifyBadEncodedMessage;
+  // Process `enc_msg` so that it equals `kSigverifyShares` if it's correct.
+  size_t i = 0;
+  for (size_t j = 0; j < kHmacDigestNumWords; ++j, ++i) {
+    enc_msg_ptr[i] ^= act_digest->digest[j] ^ kSigverifyShares[i];
   }
-  enc_msg_ptr += ARRAYSIZE(act_digest->digest);
-
   // Note: This also includes the zero byte right before PS.
   static const uint32_t kEncodedSha256[] = {
       0x05000420, 0x03040201, 0x86480165, 0x0d060960, 0x00303130,
   };
-  if (memcmp(enc_msg_ptr, kEncodedSha256, sizeof(kEncodedSha256)) != 0) {
-    return kErrorSigverifyBadEncodedMessage;
+  for (size_t j = 0; j < ARRAYSIZE(kEncodedSha256); ++j, ++i) {
+    enc_msg_ptr[i] ^= kEncodedSha256[j] ^ kSigverifyShares[i];
   }
-  enc_msg_ptr += ARRAYSIZE(kEncodedSha256);
-
   // Note: `kPsLen` excludes the last word of `enc_msg`, which is 0x0001ffff.
   static const size_t kPsLen = ARRAYSIZE(enc_msg->data) -
                                ARRAYSIZE(kEncodedSha256) -
                                ARRAYSIZE(act_digest->digest) - /*last word*/ 1;
-  uint32_t padding = UINT32_MAX;
-  for (size_t i = 0; i < kPsLen; ++i) {
-    padding &= *enc_msg_ptr++;
+  // PS up to the last word.
+  for (size_t j = 0; j < kPsLen; ++j, ++i) {
+    enc_msg_ptr[i] ^= 0xffffffff ^ kSigverifyShares[i];
+  }
+  // Last word.
+  enc_msg_ptr[i] ^= 0x0001ffff ^ kSigverifyShares[i];
+  // TODO(#10007): Use SHUTDOWN_CHECK to check loop completion.
+
+  // Reduce `enc_msg` to produce the value (`flash_exec`) to write to flash_ctrl
+  // EXEC register.
+  *flash_exec = 0;
+  uint32_t diff = 0;
+  for (size_t i = 0; i < kSigVerifyRsaNumWords; ++i) {
+    // Set diff to 0xffffffff if enc_msg[i] is wrong, no change otherwise.
+    diff |=
+        enc_msg_ptr[i] ^ kSigverifyShares[i];  // Non-zero if enc_msg is wrong.
+    diff |= ~diff + 1;          // Set upper bits to 1, 0 if no bits are set.
+    diff |= ~(diff >> 31) + 1;  // Set all bits to 1 if MSB is set, 0 otherwise.
+
+    // `flash_exec` will have the correct value if all words of the `enc_msg`
+    // are correct, but cannot guarantee that it won't be the correct value
+    // otherwise. We or it with `diff` to produce `0xffffffff` if `enc_msg` is
+    // incorrect.
+    *flash_exec ^= enc_msg_ptr[i];
+    *flash_exec |= diff;
   }
-  uint32_t res = ~padding;
-  res |= *enc_msg_ptr ^ 0x0001ffff;
-  if (res != 0) {
-    return kErrorSigverifyBadEncodedMessage;
+  // TODO(#10007): Use SHUTDOWN_CHECK to check loop completion.
+
+  rom_error_t result = *flash_exec & kSigverifyResultMask;
```

Comment:
We discussed this offline: I came up with a better way to ensure that we get a nice avalanche effect from a single bit being wrong in `flash_exec`.

Pick a random 10-bit integer a and a random 11-bit integer b. Then, redefine `kSigverifyFlashExec` to be
```c
kSigverifyFlashExec = (a << 22 | b << 11 | (a ^ b ^ kErrorOk));
``` 

Then, this operation becomes
```
rom_error_t result = (*flash_exec >> 22 ^ *flash_exec >> 11 ^ *flash_exec) & kSigverifyResultMask;
```
We now incorporate every bit of `flash_exec` into computing the return value.

Reply:
I like this suggestion, thanks @mcy! I ended up implementing this slightly differently (shifting instead of masking):
`rom_error_t result = (*flash_exec << 21 ^ *flash_exec << 10 ^ *flash_exec >> 1) >> 21;`

Reply:
BTW, the unlock value is now `0xa26a38f7` and last share is `0xd25e8754`.